engine/engine.go

```diff
@@ -89,71 +83,56 @@ func Start(ctx context.Context, startOpts *Config, fn StartCallback) error {
 	}
 	solveOpts := bkclient.SolveOpt{
 		Session: []session.Attachable{
-			secretsprovider.NewSecretProvider(secretStore),
 			socketProviders,
 			authprovider.NewDockerAuthProvider(os.Stderr),
 		},
 	}
-	if startOpts.LocalDirs == nil {
-		startOpts.LocalDirs = map[string]string{}
-	}
-	// make workdir ID unique by absolute path to prevent concurrent runs from
-	// interfering with each other
-	workdirID := fmt.Sprintf("%s_%s", workdirID, startOpts.Workdir)
-	startOpts.LocalDirs[workdirID] = startOpts.Workdir
-	solveOpts.LocalDirs = startOpts.LocalDirs
 
 	ch := make(chan *bkclient.SolveStatus)
+
+	session := core.NewSession(c, solveOpts, ch)
+
 	eg, ctx := errgroup.WithContext(ctx)
 	eg.Go(func() error {
-		var err error
-		_, err = c.Build(ctx, solveOpts, "", func(ctx context.Context, gw bkgw.Client) (*bkgw.Result, error) {
-			// Secret store is a circular dependency, since it needs to resolve
-			// SecretIDs using the gateway, we don't have a gateway until we call
-			// Build, which needs SolveOpts, which needs to contain the secret store.
-			//
-			// Thankfully we can just yeet the gateway into the store.
-			secretStore.SetGateway(gw)
-
-			coreAPI, err := schema.New(schema.InitializeArgs{
-				Router:        router,
-				SSHAuthSockID: sshAuthSockID,
-				WorkdirID:     core.HostDirectoryID(workdirID),
-				Gateway:       gw,
-				BKClient:      c,
-				SolveOpts:     solveOpts,
-				SolveCh:       ch,
-				Platform:      *platform,
-			})
-			if err != nil {
-				return nil, err
-			}
-			if err := router.Add(coreAPI); err != nil {
-				return nil, err
-			}
+		defer func() {
+			session.Wait()
+			close(ch)
```

Comment:
This makes it so we wait on any/all `mirrorCh`s to finish and then finally stop the progress UI. If we close this without waiting it'll panic (send on closed channel).